### PR TITLE
Cuda info update

### PR DESCRIPTION
**Pull request by andyneff**
- Small PR to update the info in cuda_info
- Break up large functions into smaller manageable functions
- Unit tests for new structure

### CODE REVIEW

File: linux/cuda_info.bsh [hunk 1]
```diff
@@ -235,79 +398,15 @@ function discover_cuda_info()
...
-#**
-
 #**
 # .. function:: cuda_capabilities
 #
-# :Output: * :envvar:`CUDA_ARCHES` - Supported ``CUDA`` virtual architectures (compute_xx)
-#          * :envvar:`CUDA_CODES` - Supported ``CUDA`` real architectures (sm_xx)
-#          * :envvar:`CUDA_MINIMAL_DRIVER_VERSION` - Minimal NVIDIA driver version needed for :envvar:`CUDA_VERSION` version of graphics-card driver
-#          * [:envvar:`CUDA_DEPRECATED`] - List of deprecated (yet working) architectures.
+# :Parameters: ``$1`` - The cuda version to check against. This is typically the ``nvcc`` version, but the docs version should work too. In a pinch, CUDA version will work, to some extent.
```
Comment (scott-vsi) [inline linux/cuda_info.bsh:404]:
```suggestion
# :Parameters: ``$1`` - The CUDA version to check against. This is typically the ``nvcc`` version, but the docs version should work too. In a pinch, CUDA version will work, to some extent.
```

File: linux/cuda_info.bsh [hunk 2]
```diff
@@ -22,30 +23,116 @@ source "${VSI_COMMON_DIR}/linux/aliases.bsh"
...
 #
-# :Output: ``CUDA_VERSIONS`` - List of ``CUDA`` versions found
+# :Output: :var:`CUDA_VERSIONS`
+#          :var:`CUDA_VERSION`
 #
-# Find ``CUDA`` development kits
+# Find CUDA development kits
 #
 # .. note::
 #   Will not work on macosx if mac have nvidia and two or more versions of cuda
```
[the remark this answers is not in the document]
Comment (scott-vsi) [reply linux/cuda_info.bsh:138]: (if you accept that change, delete the next line (multi-line suggestion wasn't working). 

Does docker even support the nvidia runtime on macOS?

File: tests/test-cuda_info.bsh [hunk 3]
```diff
@@ -82,7 +953,199 @@ begin_test "No Cuda on Darwin"
...
+  nvidia_docker_plugin_cuda_capability
+  [[ $(cat "${TESTDIR}/args.txt") = http://foo.bar:1234/gpu/info* ]]
+)
+end_test
+
+begin_test "Use version.txt to get version"
+(
+  setup_test
+
+  # Out of order example
```
Comment (scott-vsi) [inline tests/test-cuda_info.bsh:1030]:
```suggestion
  # Out-of-order example
```

File: tests/test-cuda_info.bsh [hunk 4]
```diff
@@ -82,7 +953,199 @@ begin_test "No Cuda on Darwin"
...
+  cuda_arch_to_cuda_family
+  assert_array_values CUDA_CARD_FAMILIES Pascal Kepler
+
+  CUDA_CARD_ARCHES=(6.1 6.1 8.0)
+  cuda_arch_to_cuda_family
+  assert_array_values CUDA_CARD_FAMILIES Pascal Pascal Ampere
+)
+end_test
+
+begin_test "Discover Cuda Info"
```
Comment (scott-vsi) [inline tests/test-cuda_info.bsh:1108]:
```suggestion
begin_test "Discover cuda info"
```

File: linux/cuda_info.bsh [hunk 5]
```diff
@@ -54,179 +141,255 @@ source "${VSI_COMMON_DIR}/linux/aliases.bsh"
...
+# .. function:: version_txt_cuda_version
 #
-# List of ``CUDA`` card architectures
+# Parses the ``version.txt`` file left behind by some CUDA installs (Version 5.5 through 11.0. 11.0 only installs as part of the "docs" package).
 #
-# An array of all the architectures the ``CUDA`` cards discovered. The only way to manually override this would be to disable :envvar:`CUDA_DISCOVER`, but that was not the intended use.
+# For versions of CUDA that are patched, the highest version number is returned.
+#
+# :Parameters: [``$1``] - Filename of ``version.txt`` file. Default: Read from stdin
 #**
```
Comment (scott-vsi) [inline linux/cuda_info.bsh:256]:
```suggestion
```

File: linux/cuda_info.bsh [hunk 6]
```diff
@@ -54,179 +141,255 @@ source "${VSI_COMMON_DIR}/linux/aliases.bsh"
...
+            fam=family[ver]
+            break
+          }
+        }
+        print fam
+      }
+    }'))
+}
+
+#**
```
Comment (scott-vsi) [inline linux/cuda_info.bsh:355]: I would move this documentation down just above the function

File: linux/cuda_info.bsh [hunk 7]
```diff
@@ -681,10 +806,9 @@ function suggested_architectures()
     # the first case first.
     if [ "${CUDA_CODES[$x]:0:1}" == "${arch:0:1}" ]; then
       CUDA_SUGGESTED_CODES+=("${CUDA_CODES[$x]}")
-    else
-      # If you get here, that means you have have a newer card than this version
-      # of cuda supports. So the best answer is to PTX compile for it.
-      CUDA_SUGGESTED_PTX+=("${CUDA_CODES[$x]}")
+    # PTX is for ARCH, not CODES
+    # else
```
Comment (scott-vsi) [inline linux/cuda_info.bsh:810]: I would move this comment into the else statement, which is allowed to pass with `:`, and then delete the `CUDA_SUGGESTED_PTX+=("${CUDA_CODES[$x]}")` line

File: linux/cuda_info.bsh [hunk 8]
```diff
@@ -22,30 +23,116 @@ source "${VSI_COMMON_DIR}/linux/aliases.bsh"
...
+# .. code-block:: bash
+#
+#     $ echo "${CUDA_ARCHES[@]}"
+#     20 30 32 35 37 50 52 53 60 61 62
+#
+#   Adding the periods to the architecture version number:
+#
+#   y=()
+#   for x in ${CUDA_ARCHES[@]+"${CUDA_ARCHES[@]}"}; do
+#     y+=("${x:0:${#x}-1}.${x:${#x}-1:1}")
```
[the remark this answers is not in the document]
Comment (andyneff) [reply linux/cuda_info.bsh:66]: So far, there is not. We'd have to see how they handle it, and add that here should that ever happens.

File: linux/cuda_info.bsh [hunk 9]
```diff
@@ -22,30 +23,116 @@ source "${VSI_COMMON_DIR}/linux/aliases.bsh"
...
+#
+# The names of the CUDA cards. Exact values may vary, e.g. "Titan X (Pascal)" vs "Titan Xp".
+#
+# ..var:: CUDA_CARD_FAMILIES
+#
+#   The family of card, for each card in :var:`CUDA_CARDS` is in.
+#
+# .. var:: CUDA_CARD_ARCHES
+#
+# The specific :var:`CUDA_ARCHES` a card natively supports, for each card in :var:`CUDA_CARDS`.
```
[the remark this answers is not in the document]
Comment (andyneff) [reply linux/cuda_info.bsh:110]: No. I thought a card only "natively" supports one arch, the rest is translated as something time.

File: linux/cuda_info.bsh [hunk 10]
```diff
@@ -54,179 +141,255 @@ source "${VSI_COMMON_DIR}/linux/aliases.bsh"
...
-#  1. Using deviceQuery (available here https://goo.gl/ocBgPU)
+# :Output: * :var:`CUDA_CARD_ARCHES`
+#          * :var:`CUDA_CARDS`
 #
-#   * looks for ${DEVICE_QUERY-deviceQuery} on the PATH
+# :Parameters: * ``DEVICE_QUERY`` - Optional path to a specific ``deviceQuery``. Default: ``deviceQuery`` (using the system path)
 #
-#  2. Using the nvidia-docker-plugin to get and parse GPU information
+# :Return Value: * ``0`` - No errors
+#                * Non-zero - If the output of ``deviceQuery`` does not contain any Cuda Capabilities, then something likely failed. This is usually caused by running a contianer without the nvidia extensions or having an insufficient nvidia driver for the CUDA running in ``deviceQuery``
```
Comment (andyneff) [inline linux/cuda_info.bsh:209]:
```suggestion
#                * Non-zero - If the output of ``deviceQuery`` does not contain any CUDA capabilities, then something likely failed. This is usually caused by running a contianer without the NVIDIA extensions or having an insufficient NVIDIA driver for the version of CUDA ``deviceQuery`` is compiled for.
```

File: linux/cuda_info.bsh [hunk 11]
```diff
@@ -54,179 +141,255 @@ source "${VSI_COMMON_DIR}/linux/aliases.bsh"
...
+  local IFS=$'\n'
+  CUDA_CARDS=($(echo "${card_info}" | \awk '/^Device/{$1=$2=""; $0=$0; $1=$1; print}' | \sed 's|"||g'))
+}
 
 #**
-# .. envvar:: CUDA_CARDS
+# .. function:: nvidia_docker_plugin_cuda_capability
 #
-# List of ``CUDA`` devices found on the computer
+# The deprecated nvidia-docker v1 API hosted GPU information, which was useful for determining the CUDA Arches. This is rarely used now.
```
[the remark this answers is not in the document]
Comment (andyneff) [reply linux/cuda_info.bsh:229]:
```suggestion
# The deprecated nvidia-docker v1 API actually hosted GPU information, which was useful for determining the CUDA Arches. This is rarely used now.
```

File: linux/cuda_info.bsh [hunk 12]
```diff
@@ -54,179 +141,255 @@ source "${VSI_COMMON_DIR}/linux/aliases.bsh"
...
-#          * :envvar:`CUDA_CARD_FAMILIES` - Matching list of ``CUDA`` families
+# Determine the version of nvcc
 #
-# Get ``CUDA`` info about each card
+# :Parameters: [``$NVCC``] - Name of nvcc executable. Default: ``nvcc``
+#
+# :Output: **stdout** - Version of ``nvcc``
 #**
-function discover_cuda_info()
+function get_nvcc_version()
```
[the remark this answers is not in the document]
Comment (andyneff) [reply linux/cuda_info.bsh:287]: Probably

File: linux/cuda_info.bsh [hunk 13]
```diff
@@ -22,211 +23,370 @@ source "${VSI_COMMON_DIR}/linux/aliases.bsh"
...
+#
+# .. var:: CUDA_CARD_ARCHES
+#
+# The specific CUDA architecture a card natively supports, for each card in :var:`CUDA_CARDS`.
+#*
+
+#*
+# .. var:: CUDA_SUGGESTED_ARCHES
+#
+# Suggested "virtual" architectures to compile for. Instead of compiling for every architecture that the :var:`CUDA_VERSION` version of CUDA supports, :var:`CUDA_SUGGESTED_ARCHES` is the intersection between :var:`CUDA_CARD_ARCHES` and :var:`CUDA_ARCHES` so that you compile only for your cards.
```
Comment (scott-vsi) [inline linux/cuda_info.bsh:116]: Is it a problem that CUDA_CARD_ARCHES seems to be a number like 6.1 whereas CUDA_ARCHES is 61

Comment (andyneff) [reply linux/cuda_info.bsh:116]: That is not the case

```
$ (. linux/cuda_info.bsh; discover_cuda_all; declare -p CUDA_CARD_ARCHES CUDA_ARCHES)
declare -a CUDA_CARD_ARCHES=([0]="61" [1]="61" [2]="30")
declare -a CUDA_ARCHES=([0]="35" [1]="37" [2]="50" [3]="52" [4]="53" [5]="60" [6]="61" [7]="62" [8]="70" [9]="72" [10]="75" [11]="80" [12]="86")
```

The unit test (and thus some of the code) is wrong

Comment (andyneff) [reply linux/cuda_info.bsh:116]: Well, it was inconsistent, but not wrong. It's consistent now

**Review by scott-vsi — APPROVED** [review]

**Pull request merged**